game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/StrengthAndRollCalculator.java

```diff
@@ -0,0 +1,43 @@
+package games.strategy.triplea.delegate.power.calculator;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.attachments.UnitSupportAttachment;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.collections.IntegerMap;
+
+/** Calculates the value of the dice roll and strength. */
+abstract class StrengthAndRollCalculator {
+
+  protected final AvailableSupports friendlySupportTracker;
+  protected final AvailableSupports enemySupportTracker;
+  // Keep track of the units that are providing support and whom they are providing it to
+  @Getter private final Map<Unit, IntegerMap<Unit>> supportGiven = new HashMap<>();
+
+  StrengthAndRollCalculator(
+      final AvailableSupports friendlySupport, final AvailableSupports enemySupport) {
+    this.friendlySupportTracker = friendlySupport.filter(getRuleFilter());
+    this.enemySupportTracker = enemySupport.filter(getRuleFilter());
+  }
+
+  abstract int getValue(Unit unit);
```

Comment:
It does not seem to be clear whether this is roll value, or strength value. Would it make sense to have two abstract methods, one for each? This way any caller can explicitly ask for bonus rolls, or bonus strength without having to infer the value returning based on the instance of the object.

Reply:
But calling `getRolls` from a strength calculator is incorrect.  So the caller would still need to know what the object instance is for them to know which method to call.

How about I switch `StrengthAndRollCalculator` to be composed inside of `StrengthCalculator` and `RollCalculator`?  Then each interface has its own clear method but the logic is still in one place.

Reply:
Can you explain that option a bit more? I don't think I quite follow the suggestion.

A method like `StrengthAndRollCalculator.getRolls()` seems to make some sense, I would also expect a `StrengthAndRollCAlculator.getStrength()` method to be present too. Splitting the two could make sense, though the data & concept are somewhat coupled and were combined from what was previously just a `Tuple<Int, Int>`. 

If you're willing to follow up, we can extract this to an issue and follow-up post-merge if that would make review and forward progress easier.

Reply:
Take a look at the new code.  There is now a `StrengthCalculator` and a `RollCalculator` interface.  The `StrengthCalculator` has `getStrength` and the `RollCalculator` has `getRoll`.  The implementing classes contain a reference to the `StrengthAndRollCalculator` (renamed as `UsedSupportTracker`) and use the public `giveSupport` and `getSupportGiven` methods to get the common logic.